libbpf-tools/task_detector.bpf.c

```diff
@@ -0,0 +1,264 @@
+} end_maps SEC(".maps");
+
+struct{
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, int);
+	__type(value, int);
+} trace_on_maps SEC(".maps");
+
+static inline void *get_user_info(void)
+{
+	int key = 0;
+
+	return bpf_map_lookup_elem(&user_info_maps, &key);
+}
+
+static inline void set_trace_on(int cpu)
+{
+	int key = 0;
+	int *trace_on = bpf_map_lookup_elem(&trace_on_maps, &key);
+
+	if (trace_on)
+		*trace_on = cpu + 1;
+}
+
+static inline void set_trace_off(void)
+{
+	int key = 0;
+	int *trace_on = bpf_map_lookup_elem(&trace_on_maps, &key);
+
+	if (trace_on)
+		*trace_on = 0;
+}
+
+static inline int should_trace(int cpu)
+{
+	int key = 0;
+	int *trace_on = bpf_map_lookup_elem(&trace_on_maps, &key);
+
+	return (trace_on && *trace_on == cpu + 1);
+}
+
+static inline void *get_trace(int key)
+{
+	return bpf_map_lookup_elem(&trace_info_maps, &key);
+}
+
+static inline void add_trace(struct trace_info ti)
+{
+	int orig_end, key = 0;
+	struct trace_info *tip;
+	int *start, *end;
+
+	start = bpf_map_lookup_elem(&start_maps, &key);
+	end = bpf_map_lookup_elem(&end_maps, &key);
+
+	if (!start || !end)
+		return;
+
+	orig_end = *end;
+	*end = (*end + 1);
+
+	if (*end == NR_ENTRY_MAX)
+		*end = 0;
+
+	if (*end == *start)
+		return;
+
+	//tip = get_trace(orig_end);
+	//if (!tip) {
+	//	*end = orig_end;
+	//	return;
+	//}
+
+	//memcpy(tip, &ti, sizeof(ti));
+	ti.ts = bpf_ktime_get_ns();
+
+	if (ti.type == TYPE_SYSCALL_ENTER ||
+	    ti.type == TYPE_SYSCALL_EXIT ||
+	    ti.type == TYPE_WAIT ||
+	    ti.type == TYPE_DEQUEUE)
+		bpf_get_current_comm(&ti.comm, sizeof(ti.comm));
+	bpf_map_update_elem(&trace_info_maps, &orig_end, &ti, 0);
+}
+
+SEC("tp_btf/sched_process_exit")
+int handle__sched_process_exit(u64 *ctx)
+{
+	struct task_struct *p = (void *)ctx[0];
+	struct user_info *ui = get_user_info();
+
+	if (ui && ui->pid && ui->pid == p->pid)
+		ui->exit = 1;
+
+	return 0;
+}
+
+SEC("tp_btf/sched_migrate_task")
+int handle__sched_migrate_task(u64 *ctx)
+{
+	struct task_struct *p = (void *) ctx[0];
+	int dest_cpu = (int) ctx[1];
+	struct trace_info ti = {
+		.cpu = dest_cpu,
+		.pid = p->pid,
+		.type = TYPE_MIGRATE,
+	};
+	struct user_info *ui = get_user_info();
+
+	if (!ui || !ui->pid || ui->pid != p->pid)
+		return 0;
+
+	set_trace_on(dest_cpu);
+
+	add_trace(ti);
+
+	return 0;
+}
+
+SEC("tp_btf/sched_wakeup")
+int handle__sched_wakeup(u64 *ctx)
+{
+	struct task_struct *p = (void *) ctx[0];
+	//int target_cpu = task_cpu(p);
+	struct trace_info ti = {
+		.cpu = p->wake_cpu,
+		.pid = p->pid,
+		.type = TYPE_ENQUEUE,
+	};
+	struct user_info *ui = get_user_info();
+
+	if (!ui || !ui->pid || ui->pid != p->pid)
+		return 0;
+
+	set_trace_on(p->wake_cpu);
+
+	add_trace(ti);
+
+	return 0;
+}
+
+SEC("tp_btf/sched_switch")
+int handle__sched_switch(u64 *ctx)
+{
+	struct task_struct *prev = (void *) ctx[1];
+	struct task_struct *next = (void *) ctx[2];
+	struct trace_info ti = {
+		.cpu = bpf_get_smp_processor_id(),

```

Comment:
you call this before you check for pid, so it will be a wasted work for a lot of cases. It's better to check pid, then do all the initializations. That way you spend resources only on events that are going to be useful.